### PR TITLE
feat: support ordered container start using initContainers (#291)

### LINKED ISSUE #291: [feature request] Support ordered container start using initContainers
**Is your feature request related to a problem? Please describe.**
Currently score-k8s doesn't support the `before` property for ordered container start (spec#104). In score-compose we already have this (score-compose#454), but score-k8s needs a different approach since Kubernetes has native initContainers.

**Describe the solution you'd like**
Map the `before` property to Kubernetes initContainers:

- `ready: complete` - regular `initContainers` entry (runs to completion before main containers start)
- `ready: started` - native sidecar `initContainers` with `restartPolicy: Always` (starts before main containers, keeps running)
- `ready: healthy` - return a validation error, same as score-compose K8s initContainers don't support health-based gating natively

For example, given:

```
containers:
  migrate:
    image: my-app:latest
    command: ["migrate"]
    before:
      app:
        ready: complete
  app:
    image: my-app:latest
```

The generated manifest would place migrate in spec.initContainers and app in spec.containers.

Changes would mainly go in internal/convert/workloads.go checking each container's Before field and splitting them into initContainers vs regular containers accordingly.

Describe alternatives you've considered Could also use Kubernetes lifecycle hooks or a custom controller, but initContainers is the most natural fit and is widely supported.

**Additional context**

- Spec issue: score-spec/spec#104
- score-compose implementation: score-spec/score-compose#454
- The before type definitions already exist in score-go v1.14.0+
- Native sidecar support (restartPolicy: Always on initContainers) is stable since K8s v1.33

Let me know if this approach makes sense or if there's anything I should adjust.

### PR DESCRIPTION
**Pull request by Abhishek9639**
#### Description
This PR adds support for the `before` property in score-k8s by mapping it to Kubernetes native `initContainers`, as described in #291.

Currently score-k8s doesn't handle the `before` field at all. With this change, containers that declare `before` entries are automatically classified and placed into `spec.initContainers` instead of `spec.containers` in the generated manifest.

The mapping follows the approach discussed in the issue:
- `ready: complete` → regular initContainer (runs to completion before main containers start)
- `ready: started` → sidecar initContainer with `restartPolicy: Always` (starts before main containers, keeps running)
- `ready: healthy` → rejected with a validation error, since K8s initContainers don't support health-based gating natively

Closes #291

#### What does this PR do?
- Updates `score-go` from v1.13.0 to v1.15.0 to get the `ContainerBefore` types
- Adds a `validateContainerBefore()` function that checks for self-references, unknown containers, cycles, and rejects `ready: healthy`
- Modifies the workload conversion to split containers into `initContainers` vs regular `containers` based on their `before` entries
- Adds 8 new tests covering both the conversion logic and validation edge cases

#### Types of changes
- [x] New feature (non-breaking change which adds functionality)

#### Checklist:
- [x] I've signed off with an email address that matches the commit author.

### CODE REVIEW

**Comment by Abhishek9639** [comment]
Hey @mathieu-benoit, 
This is my implementation for the ordered container start feature. I followed the approach discussed in the issue mapping `before` entries to Kubernetes `initContainers`, with `restartPolicy: Always` for sidecars (`ready: started`) and a validation error for `ready: healthy` since K8s doesn't support health-based gating on init containers.

I also added cycle detection and other validations similar to how it's done in score-compose. 
Let me know if anything needs to be adjusted.
Thanks

**Comment by mathieu-benoit** [comment]
Thanks, let's make sure we review and approve this one https://github.com/score-spec/score-compose/pull/454 first, and we'll all tackle this one then.

**Comment by Abhishek9639** [comment]
Got it, that makes sense. I’ll wait for score-spec/score-compose#454 to be reviewed and approved first, then we can proceed with this one. 
Thanks

**Comment by mathieu-benoit** [comment]
@Abhishek9639, now that https://github.com/score-spec/score-compose/pull/454 is approved and merged, anything here to update before our review?

**Comment by Abhishek9639** [comment]
Hi @mathieu-benoit, 
I've already updated score-go to the official v1.16.0 release and the branch is synced with main. 
Everything looks good from my end the PR is ready for your review whenever you are.
Thanks

**Comment by Abhishek9639** [comment]
Hey @mathieu-benoit,
Now that score-spec/score-compose#454 is merged, I think this one should be good to pick up next. I've synced the branch with main and everything's passing on my end. Whenever you get a chance to take a look, I'm around if anything needs adjusting.
Thanks!

**Comment by Abhishek9639** [comment]
Hey @mathieu-benoit, 
Just resolved the merge conflict in internal/command/generate_test.go (it was just my new before tests and the new KYAML/format tests from main landing in the same spot, so I kept both). Branch is synced with main again and everything's passing on my end. Should be good for review whenever you get a chance, happy to adjust anything if needed. 
Thanks!

**Comment by Abhishek9639** [comment]
Hey @mathieu-benoit,

Just resolved the conflict in internal/convert/workloads_test.go. It was only a placement conflict my four `before` tests and the new TestSharedVolumeAcrossContainers from #366 were both appended at the end of the file, and git interleaved them because they share the same `manifests, err := ConvertWorkload(...)` lines. I kept both test sets whole rather than accepting the hunks, since a straight "accept both" would have spliced the two functions together.

Also double-checked that #366 and this PR compose correctly in workloads.go: appendPodVolumes runs before the init/sidecar classification, so init container volumes still get collected and deduped at the pod level.

Branch is synced with main, build/vet/gofmt clean and the full test suite is passing on my end. Ready for review whenever you get a chance.

Thanks!

File: internal/command/generate_test.go [hunk 1]
```diff
@@ -551,6 +551,110 @@ spec:
...
+	_, _, err = executeAndResetCommand(context.Background(), rootCmd, []string{
+		"generate", "-o", "manifests.yaml", "--", "score.yaml",
+	})
+	require.NoError(t, err)
+
+	raw, err := os.ReadFile(filepath.Join(td, "manifests.yaml"))
+	require.NoError(t, err)
+	assert.Contains(t, string(raw), "initContainers")
+	assert.Contains(t, string(raw), "name: migrate")
+}
```
Comment (chris-stephenson) [inline internal/command/generate_test.go:657]: There should be a test case for cycles as well.

File: internal/convert/workloads_test.go [hunk 2]
```diff
@@ -258,6 +259,172 @@ status: {}
...
+					require.NotNil(t, ic.RestartPolicy, "sidecar should have restartPolicy")
+					assert.Equal(t, coreV1.ContainerRestartPolicyAlways, *ic.RestartPolicy)
+				}
+			}
+			return
+		}
+	}
+	t.Fatal("no Deployment found in manifests")
+}
+
```
Comment (chris-stephenson) [inline internal/convert/workloads_test.go:427]: Should we have a test case for multiple levels of dependencies? e.g. `initA -> initB -> main` ?

**Review by chris-stephenson — CHANGES_REQUESTED** [review]
Looks good.

I would add a couple of additional tests to really check all edge cases.

**Comment by Abhishek9639** [comment]
Hey @chris-stephenson, 
Thanks for the review both points are addressed in ba5b332.

Cycles: added tests for a direct loop (svc-a -> svc-b -> svc-a) and a transitive one (svc-a -> svc-b -> svc-c -> svc-a), plus a diamond case that is *not* a cycle, to make sure the check doesn't mistake a re-visited container for a loop.

Multiple levels: this one turned out to be a real bug, good catch. Init containers were being emitted alphabetically rather than in dependency order, so a chain was generated backwards whenever the names disagreed with the before graph (e.g. zulu -> alpha -> main came out as [alpha, zulu]). Your initA -> initB -> main example would have passed by luck, since it's alphabetical either way. Fixed with a topological sort of the init containers, with ties broken alphabetically so the manifests stay deterministic. Covered by four tests, including the non-alphabetical chain and a sidecar in the middle of one.

Full suite is green. Let me know if you'd like anything adjusted.

File: internal/command/generate_test.go (hunk 1 above)
Comment (Abhishek9639) [reply internal/command/generate_test.go:657]: Done!

File: internal/convert/workloads_test.go (hunk 2 above)
Comment (Abhishek9639) [reply internal/convert/workloads_test.go:427]: Done!

File: internal/convert/workloads_test.go [hunk 3]
```diff
@@ -486,3 +486,196 @@ func TestSharedVolumeAcrossContainers(t *testing.T) {
...
+			assert.Len(t, dep.Spec.Template.Spec.Containers, 1, "expected 1 regular container")
+			assert.Equal(t, "app", dep.Spec.Template.Spec.Containers[0].Name)
+			return
+		}
+	}
+	t.Fatal("no Deployment found in manifests")
+}
+
+func TestConvertWorkload_BeforeIndependentInitsStayAlphabetical(t *testing.T) {
```
Comment (chris-stephenson) [inline internal/convert/workloads_test.go:632]: I'm not really sure that this test does anything. Key order in go is random, but selecting from the same map many times will return the same key order each time.

At the very least, the loop should include the generation of the map `state.Containers` and ideally you would have many more keys to increase the chance of a missordering occuring.

**Review by chris-stephenson — CHANGES_REQUESTED** [review]
One test seems to not do what it is intended to do.

File: internal/convert/workloads_test.go (hunk 3 above)
Comment (Abhishek9639) [reply internal/convert/workloads_test.go:632]: Done!

**Comment by Abhishek9639** [comment]
You're right that it wasn't testing much. Fixed both things:

- The workload is now rebuilt inside the loop, so the container map gets constructed fresh on every iteration instead of being converted over and over.
- Bumped it from 2 init containers to 12, inserted in reverse alphabetical order.

To make sure it actually catches something now, I commented out `slices.Sort(containerNames)` in `workloads.go` and re-ran it fails 5 out of 5 runs, and goes green again once the sort is back.

One small thing I ran into while looking at this, which I think supports your point rather than contradicting it. Ranging the *same* map repeatedly does give you different orders I ranged a 3-key map 1000 times and got:

```
seed migrate app      753
app seed migrate      121
migrate app seed      126
```

But that's only 3 of the 6 possible permutations. Go randomises the starting offset rather than doing a real shuffle, so iteration ends up being a rotation and some orderings just can't come out of a given map. So rebuilding the map is doing more work than re-converting would, which is what you were asking for anyway.

The old version wasn't completely dead (with 2 containers it had ~50% odds of catching a misordering each iteration), but 12 containers makes an accidental pass about 1 in 12!, which feels much safer.

**Review by chris-stephenson — APPROVED** [review]
LGTM

Thanks for the contribution @Abhishek9639 !
